OpenRA.Mods.Common/Widgets/ResourcePreviewWidget.cs

```diff
@@ -29,39 +31,43 @@ public class ResourcePreviewWidget : Widget
 
 		string resourceType;
 		IResourceRenderer resourceRenderer;
+		int2 spriteOffset;
 
-		public string ResourceType
+		public void SetResourceType(string value)
 		{
-			get => resourceType;
+			resourceType = value;
+			if (resourceType != null)
+				resourceRenderer = resourceRenderers.FirstOrDefault(r => r.ResourceTypes.Contains(resourceType));
+			else
+				resourceRenderer = null;
+
+			var bounds = resourceRenderer?.RenderUIPreview(worldRenderer, resourceType, int2.Zero, viewportSizes.DefaultScale)
+				.FirstOrDefault()
+				.PrepareRender(worldRenderer)
+				?.ScreenBounds(worldRenderer);
 
-			set
+			if (bounds != null)
 			{
-				resourceType = value;
-				if (resourceType != null)
-					resourceRenderer = resourceRenderers.FirstOrDefault(r => r.ResourceTypes.Contains(resourceType));
-				else
-					resourceRenderer = null;
+				spriteOffset = -bounds.Value.Location;
+				IdealPreviewSize = bounds.Value.Size;
 			}
+			else
+				IdealPreviewSize = new Size((int)(tileSize.Width * viewportSizes.DefaultScale), (int)(tileSize.Height * viewportSizes.DefaultScale));
 		}
 
-		public Size IdealPreviewSize { get; }
-
 		[ObjectCreator.UseCtor]
 		public ResourcePreviewWidget(ModData modData, WorldRenderer worldRenderer, World world)
 		{
 			this.worldRenderer = worldRenderer;
 			viewportSizes = modData.Manifest.Get<WorldViewportSizes>();
-			resourceRenderers = world.WorldActor.TraitsImplementing<IResourceRenderer>().ToArray();
+			resourceRenderers = [.. world.WorldActor.TraitsImplementing<IResourceRenderer>()];
```

Comment:
In a later style discussion there was a preference to avoid spread syntax for this.